File: iosApp/iosApp/Presentation/Main/Ranking/RankingViewModel.swift
//
//  RankingViewModel.swift
//  ConCafe
//
//  Created by 홍희표 on 2026/03/06.
//

import Foundation
import Combine
import Shared
import KMPNativeCoroutinesAsync

@MainActor
final class RankingViewModel: ObservableObject {
    private let getRankingFeedUseCase: GetRankingFeedUseCase

    private let cafeDetailEventPublisher: CafeDetailEventPublisher

    private let castEventPublisher: CastEventPublisher

    @Published private(set) var uiState = RankingUiState.empty

    let event = PassthroughSubject<RankingEvent, Never>()

    private var tasks: [TaskKey: Task<Void, Never>] = [:]

    private func loadRankingFeed() {
        let period = uiState.selectedPeriod
        let country = uiState.selectedRegion.country
        let city = uiState.selectedRegion.city

        uiState.isLoading = true
        uiState.errorMessage = nil
        Task {
            do {
                let result = try await getRankingFeedUseCase.invoke(period: period, country: country, city: city)

                if let success = result as? AppResultSuccess<AnyObject>,
                   let feed = success.data as? Shared.RankingFeed {
                    uiState.isLoading = false
                    uiState.errorMessage = nil
                    uiState.selectedAdIndex = 0
                    uiState.ads = feed.ads
                    uiState.maidRankings = feed.castRankings
                    uiState.cafeRankings = feed.cafeRankings
                } else {
                    uiState.isLoading = false
                    uiState.errorMessage = "랭킹 데이터를 불러오지 못했습니다."
                }
            } catch {
                uiState.isLoading = false
                uiState.errorMessage = "랭킹 데이터를 불러오지 못했습니다."
            }
        }
    }

    private func observeCafeDetailEvent() {
        tasks[.cafeDetailEvent]?.cancel()
        tasks[.cafeDetailEvent] = Task {
            do {
                for try await event in asyncSequence(for: cafeDetailEventPublisher.events) {
                    if let updated = event as? CafeDetailEvent.CafeInfoUpdated {
                        self.patchCafeRanking(updated.cafe)
                    }
                }
            } catch {
                print("Error: \(error)")
            }
        }
    }

    private func observeCastEvent() {
        tasks[.castEvent]?.cancel()
        tasks[.castEvent] = Task {
            do {
                for try await event in asyncSequence(for: castEventPublisher.events) {
                    switch event {
                    case let updated as Shared.CastEvent.Updated:
                        self.patchCastRanking(updated.cast)
                    case let deleted as Shared.CastEvent.Deleted:
                        self.uiState.maidRankings.removeAll { $0.id == deleted.castId }
                    default:
                        break
                    }
                }
            } catch {
                print("Error: \(error)")
            }
        }
    }

    private func patchCafeRanking(_ cafe: Cafe) {
        let subtitle = cafe.region.address.components(separatedBy: "구").first?
            .components(separatedBy: "로").first?
            .trimmingCharacters(in: .whitespacesAndNewlines)
        let nextSubtitle = (subtitle?.isEmpty == false ? subtitle! : cafe.region.city)
        uiState.cafeRankings = uiState.cafeRankings.map { entry in
            guard entry.id == cafe.id else { return entry }
            return RankingFeedEntry(
                id: entry.id,
                rank: entry.rank,
                name: cafe.name,
                subtitle: nextSubtitle,
                score: entry.score,
                change: entry.change,
                startColorHex: entry.startColorHex,
                endColorHex: entry.endColorHex,
                symbol: entry.symbol
            )
        }
    }

    private func patchCastRanking(_ cast: Cast) {
        uiState.maidRankings = uiState.maidRankings.map { entry in
            guard entry.id == cast.id else { return entry }
            return RankingFeedEntry(
                id: entry.id,
                rank: entry.rank,
                name: cast.name,
                subtitle: entry.subtitle,
                score: entry.score,
                change: entry.change,
                startColorHex: entry.startColorHex,
                endColorHex: entry.endColorHex,
                symbol: entry.symbol
            )
        }
    }

    func onAction(_ action: RankingAction) {
        switch action {
        case .changeTab(let tab):
            uiState.selectedTab = tab
        case .changePeriod(let period):
            uiState.selectedPeriod = period
            loadRankingFeed()
        case .changeRegion(let region):
            uiState.selectedRegion = region
            loadRankingFeed()
        case .selectAd(let index):
            let lastIndex = max(uiState.ads.count - 1, 0)
            uiState.selectedAdIndex = min(max(index, 0), lastIndex)
        case .tapMaid(let id):
            event.send(.navigateToCast(id: id))
        case .tapCafe(let id):
            event.send(.navigateToCafe(id: id))
        }
    }

    init(
        getRankingFeedUseCase: GetRankingFeedUseCase = KoinInitializerKt.resolveGetRankingFeedUseCase(),
        cafeDetailEventPublisher: CafeDetailEventPublisher = KoinInitializerKt.resolveCafeDetailEventPublisher(),
        castEventPublisher: CastEventPublisher = KoinInitializerKt.resolveCastEventPublisher()
    ) {
        self.getRankingFeedUseCase = getRankingFeedUseCase
        self.cafeDetailEventPublisher = cafeDetailEventPublisher
        self.castEventPublisher = castEventPublisher
        
        observeCafeDetailEvent()
        observeCastEvent()
        loadRankingFeed()
    }

    deinit {
        tasks.values.forEach { $0.cancel() }
        tasks.removeAll()
    }

    private enum TaskKey {
        case cafeDetailEvent
        case castEvent
    }
}

File: composeApp/src/commonMain/kotlin/com/hhp227/concafe/presentation/main/ranking/RankingViewModel.kt
package com.hhp227.concafe.presentation.main.ranking

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.hhp227.concafe.domain.common.AppResult
import com.hhp227.concafe.domain.model.Cafe
import com.hhp227.concafe.domain.event.CafeDetailEvent
import com.hhp227.concafe.domain.model.Cast
import com.hhp227.concafe.domain.event.CastEvent
import com.hhp227.concafe.domain.event.publisher.CafeDetailEventPublisher
import com.hhp227.concafe.domain.event.publisher.CastEventPublisher
import com.hhp227.concafe.domain.usecase.ClearNativeAdUseCase
import com.hhp227.concafe.domain.usecase.GetRankingFeedUseCase
import com.hhp227.concafe.domain.usecase.LoadNativeAdUseCase
import com.hhp227.concafe.domain.usecase.ObserveCurrentUserUseCase
import com.hhp227.concafe.presentation.main.ranking.RankingEvent.*
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch

class RankingViewModel(
    private val getRankingFeedUseCase: GetRankingFeedUseCase,
    private val observeCurrentUserUseCase: ObserveCurrentUserUseCase,
    private val loadNativeAdUseCase: LoadNativeAdUseCase,
    private val clearNativeAdUseCase: ClearNativeAdUseCase,
    private val cafeDetailEventPublisher: CafeDetailEventPublisher,
    private val castEventPublisher: CastEventPublisher
) : ViewModel() {
    private val _uiState = MutableStateFlow(RankingUiState.empty)
    val uiState = _uiState.asStateFlow()

    private val _event = MutableSharedFlow<RankingEvent>()
    val event = _event.asSharedFlow()

    private val jobs = mutableMapOf<TaskKey, Job>()

    private fun loadRankingFeed() {
        val currentState = _uiState.value
        val country = currentState.selectedRegion.country
        val city = currentState.selectedRegion.city
        val period = currentState.selectedPeriod

        _uiState.update { it.copy(isLoading = true, errorMessage = null) }

        viewModelScope.launch {
            when (val result = getRankingFeedUseCase.invoke(period, country, city)) {
                is AppResult.Success -> {
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            errorMessage = null,
                            selectedAdIndex = 0,
                            ads = result.data.ads,
                            maidRankings = result.data.castRankings,
                            cafeRankings = result.data.cafeRankings
                        )
                    }
                }
                is AppResult.Failure -> {
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            errorMessage = "랭킹 데이터를 불러오지 못했습니다."
                        )
                    }
                }
            }
        }
    }

    private fun observeCafeDetailEvent() {
        jobs[TaskKey.OBSERVE_CAFE_DETAIL_EVENT]?.cancel()
        jobs[TaskKey.OBSERVE_CAFE_DETAIL_EVENT] = viewModelScope.launch {
            cafeDetailEventPublisher.events.collectLatest { event ->
                if (event is CafeDetailEvent.CafeInfoUpdated) {
                    patchCafeRanking(event.cafe)
                }
            }
        }
    }

    private fun observeCastEvent() {
        jobs[TaskKey.OBSERVE_CAST_EVENT]?.cancel()
        jobs[TaskKey.OBSERVE_CAST_EVENT] = viewModelScope.launch {
            castEventPublisher.events.collectLatest { event ->
                when (event) {
                    is CastEvent.Created -> Unit
                    is CastEvent.Updated -> patchCastRanking(event.cast)
                    is CastEvent.Deleted -> removeCastRanking(event.castId)
                }
            }
        }
    }

    private fun observeSession() {
        jobs[TaskKey.OBSERVE_SESSION]?.cancel()
        jobs[TaskKey.OBSERVE_SESSION] = viewModelScope.launch {
            observeCurrentUserUseCase.invoke().collectLatest { user ->
                _uiState.update {
                    it.copy(
                        isLoggedIn = user != null,
                        isLoginPromptVisible = if (user == null) it.isLoginPromptVisible else false
                    )
                }
            }
        }
    }

    private fun patchCafeRanking(cafe: Cafe) {
        val subtitle = cafe.region.address.substringBefore("구").substringBefore("로").ifBlank { cafe.region.city }
        _uiState.update { state ->
            state.copy(
                cafeRankings = state.cafeRankings.map { entry ->
                    if (entry.id == cafe.id) {
                        entry.copy(name = cafe.name, subtitle = subtitle)
                    } else {
                        entry
                    }
                }
            )
        }
    }

    private fun patchCastRanking(cast: Cast) {
        _uiState.update { state ->
            state.copy(
                maidRankings = state.maidRankings.map { entry ->
                    if (entry.id == cast.id) {
                        entry.copy(name = cast.name)
                    } else {
                        entry
                    }
                }
            )
        }
    }

    private fun removeCastRanking(castId: String) {
        _uiState.update { state ->
            state.copy(maidRankings = state.maidRankings.filterNot { it.id == castId })
        }
    }

    fun loadNativeAdIfNeeded() {
        if (_uiState.value.nativeAd != null) return
        viewModelScope.launch {
            val ad = loadNativeAdUseCase.invoke()

            _uiState.update {
                it.copy(nativeAd = ad)
            }
        }
    }

    fun clearAd() {
        clearNativeAdUseCase.invoke(uiState.value.nativeAd)
        _uiState.update {
            it.copy(nativeAd = null)
        }
    }

    fun onAction(action: RankingAction) {
        when (action) {
            is RankingAction.ChangePeriod -> {
                _uiState.update { it.copy(selectedPeriod = action.period) }
                loadRankingFeed()
            }
            is RankingAction.ChangeRegion -> {
                _uiState.update { it.copy(selectedRegion = action.region) }
                loadRankingFeed()
            }
            is RankingAction.ChangeTab -> _uiState.update { it.copy(selectedTab = action.tab) }
            is RankingAction.SelectAd -> _uiState.update {
                val lastIndex = (it.ads.size - 1).coerceAtLeast(0)
                it.copy(selectedAdIndex = action.index.coerceIn(0, lastIndex))
            }
            is RankingAction.ClickMaid -> viewModelScope.launch {
                requireSignedIn { _event.emit(NavigateToCast(action.id)) }
            }
            is RankingAction.ClickCafe -> viewModelScope.launch {
                requireSignedIn { _event.emit(NavigateToCafe(action.id)) }
            }
            RankingAction.ClickLoginPromptSignIn -> viewModelScope.launch {
                _uiState.update { it.copy(isLoginPromptVisible = false) }
                _event.emit(NavigateToSignIn)
            }
            RankingAction.DismissLoginPrompt -> _uiState.update { it.copy(isLoginPromptVisible = false) }
            is RankingAction.UpdateBannerHeight -> {
                _uiState.update { state ->
                    if (action.heightPx > state.bannerHeightPx) {
                        state.copy(bannerHeightPx = action.heightPx)
                    } else {
                        state
                    }
                }
            }
        }
    }

    private suspend fun requireSignedIn(onAuthenticated: suspend () -> Unit) {
        if (_uiState.value.isLoggedIn) {
            onAuthenticated()
        } else {
            _uiState.update { it.copy(isLoginPromptVisible = true) }
        }
    }

    override fun onCleared() {
        jobs.values.forEach(Job::cancel)
        jobs.clear()
        clearAd()
        super.onCleared()
    }

    init {
        observeSession()
        observeCafeDetailEvent()
        observeCastEvent()
        loadRankingFeed()
        loadNativeAdIfNeeded()
    }

    private enum class TaskKey {
        OBSERVE_CAFE_DETAIL_EVENT,
        OBSERVE_CAST_EVENT,
        OBSERVE_SESSION
    }
}
